### PR TITLE
Newznab to YML Script & Create v8 Definitions

### PR DESCRIPTION
**Pull request by bakerboy448**
ToDos:
- [x] - Remove Duplicates for Search Inputs  (Baker: Best way to do so?)
```
MethodInvocationException: C:\Development\Code\Prowlarr_Indexers\scripts\NewznabToCardigannYML.ps1:152
Line |
 152 |          $inputs.Add($($searchinput), "{{ .Query.$($searchstring)}}")
     |                                                    ~~~~~~~~~~~~~
     | Exception calling "Add" with "2" argument(s): "Item has already been added. Key in dictionary:
     | 'traktid'  Key being added: 'traktid'"
```
- [x] - Do not create YML if response is not 200 (redirect if needed)
- [x] - Review Code/Structure/Baker's Changes (@RobinDadswell)
- [x] - Fix Indents on: (not necessary)
  - [x] sites block (url) URL
  - [x] settings block
  - [x] search error block
  - [x] search paths block
  - [X] download error block
- [x] - Regenerate definitions after fixing indents
  - Powershell commands are [here](https://docs.google.com/spreadsheets/d/e/2PACX-1vQBgkgN8XUlsBS83ogM1l3uGIBOY5iZPdrl58cSYyro2hgUDLH-8cGLr_bqdvo4rRnHvAx8-VHfCoFI/pub?output=xlsx) (XLSX warning) - on the `ScriptInputs` worksheet
- [x] - Manually or automatically validate categories and comment out invalid ones
- [x] - Catch no XML returned 
image
- [x] - dupe query examples 
image
- [x] - Better Error Code Output if not 200
image
- [x] - Fix better category validation
```
commented out XXX/SD

yet
6080,XXX/SD on the list
same for 
+#  - { id: 1, cat: Other, desc: Other }
+#  - { id: 10, cat: Other/Misc, desc: Other/Misc }
+#  - { id: 20, cat: Other/Hashed, desc: Other/Hashed }

yet Other is fine
```

### CODE REVIEW

**Comment by RobinDadswell** [comment]
Why are we trying to fix the indenting? If the hashtable has correct indenting then the YML indenting is correct...

**Comment by bakerboy448** [comment]
> Why are we trying to fix the indenting? If the hashtable has correct indenting then the YML indenting is correct... 

Thought Cardigann was very particular about the indenting?

Also there's several areas where the indenting outright isn't occurring

File: definitions/v4/drunkenslug.yml [hunk 1]
```diff
@@ -180,8 +166,8 @@ search:
...
 download:
   error:
-    - selector: error
-      message:
-        selector: error
-        attribute: description
+  - selector: error
+    message:
+      selector: error
+      attribute: description
```
Comment (RobinDadswell) [inline definitions/v4/drunkenslug.yml:172]: @bakerboy448  this is valid YML as is the previous - as long as Cardigann doesn't complain (@Qstick are you able to confirm?) then we should be fine

File: scripts/NewznabToCardigannYML.ps1 [hunk 2]
```diff
@@ -35,208 +48,273 @@ function Invoke-NameReplace {
...
+    foreach ($subcategory in $category.subcat)
+    {
+        $subcatName = Invoke-CatNameReplace -Name "$($catName)/$($subcategory.name)"
+        $subtemp = [PSCustomObject][ordered]@{
+            id   = $subcategory.id
+            cat  = $subcatName
+            desc = "'$($catName)/$($subcategory.name)'"
+        }
+        $categories.Add($subtemp)
     }
```
Comment (RobinDadswell) [inline scripts/NewznabToCardigannYML.ps1:105]: If merging these two you may as well get rid of the $ymlcategories bit and do it all here

File: scripts/NewznabToCardigannYML.ps1 [hunk 3]
```diff
@@ -35,208 +48,273 @@ function Invoke-NameReplace {
...
+    q      = '{{ .Keywords }}'
+}
+
+foreach ($searchinput in ($modes.GETENUMERATOR() | ForEach-Object { $_.VALUE }).Replace('q', '').Replace('[', '').Replace(']', '').Split(',') | Sort-Object)
+{
+    if ($searchinput -ne 'q' -and $searchinput -ne '')
+    {
+        #return $searchinput
+        [string]$searchstring = $(Invoke-ModesReplace -Name $($searchinput))
+        $inputs.Add($($searchinput), "{{ .Query.$($searchstring)}}")
```
Comment (RobinDadswell) [inline scripts/NewznabToCardigannYML.ps1:152]: You have said this is getting some elements twice - just do an if inputs doesn't contain on the key

**Review by RobinDadswell — CHANGES_REQUESTED** [review]

File: scripts/NewznabToCardigannYML.ps1 [hunk 4]
```diff
@@ -7,62 +9,312 @@
...
+    param (
+        [Parameter(Mandatory)]
+        [string]
+        $Name
+    )
+    return ($Name -replace 'category2', 'category' -replace '"{ ', '{ ' -replace ' }"', ' }')
+}
 # Get Data and digest objects
+Write-Information 'Requesting Caps'
 [xml]$xmlResponse = (Invoke-WebRequest -Uri $capscall -Headers $headers -ContentType 'application/xml' -Method Get).Content
```
Comment (RobinDadswell) [inline scripts/NewznabToCardigannYML.ps1:71]: headers is never defined

**Comment by RobinDadswell** [comment]
> * - Remove Duplicates for Search Inputs  (Baker: Best way to do so?)

Tried two or 3 indexers and could not replicate - let me know which one and I will fix

File: scripts/NewznabToCardigannYML.ps1 (hunk 3 above)
Comment (bakerboy448) [reply scripts/NewznabToCardigannYML.ps1:139]: > C:\Development\Code\Prowlarr_Indexers\scripts\NewznabToCardigannYML.ps1 -site "https://www.tabula-rasa.pw" -indexer "Tabula Rasa" -privacy "private" -apipath "/api/v1/api"-outputfile "C:\Development\Code\Prowlarr_Indexers\definitions\v4\tabularasa.yml"

File: scripts/NewznabToCardigannYML.ps1 [hunk 5]
```diff
@@ -7,62 +9,329 @@
...
+$categoryCsv = Import-Csv $($PSScriptRoot + [system.Io.Path]::DirectorySeparatorChar + 'newznabcats(v4.0).csv')
+[System.Collections.Generic.List[string]]$linesToReplace = @()
+$ymlout = Get-Content $OutputFile -Encoding utf8
+#Validating Category Names
+foreach ($line in ($ymlout | Select-String '{ id:' | Select-Object -ExpandProperty Line))
+{
+    foreach ($cleanLine in ($line -replace '- { ', '' -replace ' }' -replace 'id: ', '' -replace 'cat: ').Trim())
+    {
+        $split = $cleanLine -split ', '
+        $categoryCsvCategory = ($categoryCsv | Where-Object { $_.id -eq $split[0] }).newznabcat
```
Comment (bakerboy448) [inline scripts/NewznabToCardigannYML.ps1:320]: we cannot use ID to match.  one of the big purposes of switching to YML is to allow remapping of a site's cat # to the correct newznab cat as they do not follow standards

**Comment by bakerboy448** [comment]
🥳 - Script Works
👍 - Output is Sane

@Qstick - Please review the YML outputs and Replace in the Script for Category Mapping
@RobinDadswell - Please review the script and ensure it seems sane.

The following Categories are not mapped and thus are commented out:
```
WARNING: YML Category of Indexer ID [4090] parsed as Category [PC/Bookware] from Indexer's Category Name [PC/Bookware] is not Newznab Standard. Category disabled in YML
WARNING: YML Category of Indexer ID [1030] parsed as Category [Console/Nintendo] from Indexer's Category Name [Console/Nintendo] is not Newznab Standard. Category disabled in YML
WARNING: YML Category of Indexer ID [100003] parsed as Category [Console/PS2] from Indexer's Category Name [Console/PS2] is not Newznab Standard. Category disabled in YML
WARNING: YML Category of Indexer ID [100006] parsed as Category [Console/PS1] from Indexer's Category Name [Console/PS1] is not Newznab Standard. Category disabled in YML
WARNING: YML Category of Indexer ID [100007] parsed as Category [Console/NSwitch] from Indexer's Category Name [Console/NSwitch] is not Newznab Standard. Category disabled in YML
WARNING: YML Category of Indexer ID [4080] parsed as Category [PC/3dModels] from Indexer's Category Name [PC/3dModels] is not Newznab Standard. Category disabled in YML
WARNING: YML Category of Indexer ID [1110] parsed as Category [Console/NSW] from Indexer's Category Name [Console/NSW] is not Newznab Standard. Category disabled in YML
WARNING: YML Category of Indexer ID [4090] parsed as Category [PC/Bookware] from Indexer's Category Name [PC/Bookware] is not Newznab Standard. Category disabled in YML
WARNING: YML Category of Indexer ID [90000] parsed as Category [Suspect] from Indexer's Category Name [Suspect] is not Newznab Standard. Category disabled in YML
WARNING: YML Category of Indexer ID [90010] parsed as Category [Suspect/Filtered] from Indexer's Category Name [Suspect/Filtered] is not Newznab Standard. Category disabled in YML
WARNING: YML Category of Indexer ID [1096] parsed as Category [Console/NSwitch] from Indexer's Category Name [Console/NSwitch] is not Newznab Standard. Category disabled in YML
```

Also note that no definition can be generated for Usenet Crawler
```
The status code [504] was received from the website [https://www.usenet-crawler.com/api?t=caps], please investigate why and try again when issue is resolved
```

File: scripts/NewznabToCardigannYML.ps1 [hunk 6]
```diff
@@ -7,62 +9,445 @@
...
+    [string]$indexer,
+    [Parameter(Position = 2)]
+    [string]$privacy,
     [Parameter(Position = 3)]
-    [string]$indexer
+    [string]$apipath = '/api',
+    [Parameter(Position = 4)]
+    [System.IO.FileInfo]$outputfile = ".$([System.IO.Path]::DirectorySeparatorChar)newznab.yml",
+    [Parameter(Position = 4)]
+    [string]$language = 'en-US'
```
Comment (RobinDadswell) [inline scripts/NewznabToCardigannYML.ps1:34]: Double position 4?

**Review by RobinDadswell — CHANGES_REQUESTED** [review]
Double 4  position and then squash this down to a commit with all the changes for the script and then another commit for the definitions

**Comment by bakerboy448** [comment]
> Double 4 position and then squash this down to a commit with all the changes for the script and then another commit for the definitions

good to go

**Comment by RobinDadswell** [comment]
As long as @Qstick  is happy with the yml, I am happy with the script

File: definitions/v5/rutracker.yml [hunk 7]
```diff
@@ -609,7 +609,7 @@ caps:
   - { id: 899, cat: PC/Games, desc: "Игры для Linux" }
   - { id: 537, cat: PC/Mac, desc: "|- Нативные игры для Mac" }
   - { id: 900, cat: PC/Games, desc: "|- Визуальные новеллы" }
-  - { id: 1008, cat: PC/Games, desc: '|- Квесты в стиле "Поиск предметов"' }
+  - { id: 1008, cat: PC/Games, desc: "|- Квесты в стиле "Поиск предметов"" }
```
Comment (bakerboy448) [inline definitions/v5/rutracker.yml:612]:
```suggestion
  - { id: 1008, cat: PC/Games, desc: '|- Квесты в стиле "Поиск предметов"' }
```

**Comment by Qstick** [comment]
Conflict, Also should these be v6? Also need to add SceneHD

**Comment by bakerboy448** [comment]
> Conflict, Also should these be v6? Also need to add SceneHD

- [ ] - Add SceneHD

**Comment by Qstick** [comment]
What happened to all the other ymls?

**Comment by bakerboy448** [comment]
> What happened to all the other ymls?

Probably lost in the rebase with the dupe names and git rename

Will need to try to hunt through reflog

**Comment by Qstick** [comment]
> @Qstick can't find in the reflog; even before v5 => v6 so they're lost in the abyss :/

Killing me, that was 50 ymls.

**Comment by bakerboy448** [comment]
> > @Qstick can't find in the reflog; even before v5 => v6 so they're lost in the abyss :/
> 
> Killing me, that was 50 ymls.

Fixed - Thank you Git reflog 5a0a71117e1b6960f5450c1e0084176775a463be

**Comment by bakerboy448** [comment]
pushed the `best` commit I can find. which does not seem to have implementation for C# torrents; only newznab

`git reflog show Newznab-xml-to-yml-scriptupdates --date=iso --pretty=short`
image

**Comment by bakerboy448** [comment]
need to bump to v8

**Comment by Qstick** [comment]
Updated with all the YMLs for C# Indexers and rebased

**Pull request closed without merging**